### ISSUE TITLE
Backward movement functions do not work in forms within templates

### ISSUE THREAD

**Issue by villarragut** [minad/tempel#127]
When you insert a form like `(backward-char)`, `(search-backward "a")`, or anything moving the point backward in a template, it does not work. The forward versions of these functions work perfectly.

Thanks for the good work!

Best wishes,
Victor

**Comment by minad** [comment]
I cannot reproduce this, for example this template works for me as expected, resulting in the string `fobaro`.

~~~
(backward "foo" (backward-char) "bar")
~~~

**Comment by villarragut** [comment]
Thanks for your response. Indeed, your template works as expected for me too.

Here's an explanation of what I'm trying to achieve with Tempel.

I'm using the package Xenops for LaTeX. When editing an image generated from a formula between dollars, it always places the cursor right after the first dollar. E.g. if the formula is
`$e^{i\pi}+1=0$`
then the cursor is between the first dollar and the e initially.

I'm using Tempel to insert a template consisting of $$ with the cursor in between. When the cursor is already in a math environment (between two dollars), I'd like the same template to:

1. if the cursor is between the 0 and the second $ in the example, then move to the position between the first dollar and the e;
2. if the cursor is anywhere else, then move it to the position between the 0 and the second $.

In order to get this, I'm doing the following.

**TL;DR**

I defined the following templates:
`latex-mode :when (texmathp)
(math-inline  (progn (cond ((eq (char-after) ?$) (search-backward "$") (forward-char)) (t (search-forward "$") (backward-char))) ""))
latex-mode :when (not (texmathp))
(math-inline "$" q "$")`

Then I just set up a key binding like this:
`(tempel-key "$" math-inline LaTeX-mode-map)`

I'm sure I'm making a mistake somewhere, but I can't see where.

Thanks again.

**Comment by minad** [comment]
This sounds a bit too complicated for me to investigate. Tempel uses a simple interpreter loop which evaluates each expression step by step, see https://github.com/minad/tempel/blob/38dad1f85104c7c734ad9245f3a85b49562332d0/tempel.el#L340-L371. The behavior should be quite predictable.

**Comment by villarragut** [comment]
It certainly helps me understand how Tempel works.

I think I can achieve my goal just by using an Elisp function, which probably is the reasonable way to go if I want to move the point instead of inserting a template anyway!

Thank you for your time, your advice, and your wonderful work.

**Comment by minad** [comment]
> In short: Snippet packages (unless I have missed a more sophisticated and advanced implementation somewhere) as they are parsing/loading/expanding the snippet: for each special character, variable, syntax, etc. they immediately make buffer modifications. Any cursor based methods or manipulations virtually impossible. The cursor (i.e. point) is not stable during that process.

Thanks, exactly! This is by design. A template is a sequence of operations, executed, as if you would type them out by hand or execute a macro. I tried to say this above, but probably a little bit cryptic: "Tempel uses a simple interpreter loop which evaluates each expression step by step".

**Comment by villarragut** [comment]
Thank you for your thorough response, Matt.

Indeed, it makes it easier to understand what puzzled me in the first place and why the implementation is the way it is.
